api/src/main/java/org/eclipse/microprofile/metrics/MetricRegistry.java

```diff
@@ -625,72 +918,119 @@ public static String name(Class<?> klass, String... names) {
      *
      * @param filter the metric filter to match
      * @return all the timers in the registry
      */
-    public abstract SortedMap<MetricID, Timer> getTimers(MetricFilter filter);
+    default SortedMap<MetricID, Timer> getTimers(MetricFilter filter) {
+        return getMetrics(Timer.class, filter);
+    }
 
     /**
      * Returns a map of all the simple timers in the registry and their {@link MetricID}s.
      *
      * @return all the timers in the registry
      */
-    public abstract SortedMap<MetricID, SimpleTimer> getSimpleTimers();
+    SortedMap<MetricID, SimpleTimer> getSimpleTimers();
 
     /**
      * Returns a map of all the simple timers in the registry and their {@link MetricID}s which match the given filter.
      *
      * @param filter the metric filter to match
      * @return all the timers in the registry
      */
-    public abstract SortedMap<MetricID, SimpleTimer> getSimpleTimers(MetricFilter filter);
+    default SortedMap<MetricID, SimpleTimer> getSimpleTimers(MetricFilter filter) {
+        return getMetrics(SimpleTimer.class, filter);
+    }
+
+    /**
+     * Returns a map of all the metrics in the registry and their {@link MetricID}s which match the given filter.
+     *
+     * @param filter the metric filter to match
+     * @return all the metrics in the registry
+     *
+     * @since 2.4
+     */
+    SortedMap<MetricID, Metric> getMetrics(MetricFilter filter);
+
+    /**
+     * Returns a map of all the metrics in the registry and their {@link MetricID}s which match the given filter
+     * and which are assignable to the provided type.
+     * @param ofType the type to which all returned metrics should be assignable
+     * @param filter the metric filter to match
+     *
+     * @return all the metrics in the registry
+     *
+     * @since 2.4
+     */
+    @SuppressWarnings("unchecked")
+    default <T extends Metric> SortedMap<MetricID, T> getMetrics(Class<T> ofType, MetricFilter filter) {
+        return (SortedMap<MetricID, T>) getMetrics(
+                (metricID, metric) -> filter.matches(metricID, metric)
+                    && ofType.isAssignableFrom(metric.getClass()));
+    }
 
     /**
      * Returns a map of all the metrics in the registry and their {@link MetricID}s.
      *
      * @return all the metrics in the registry
+     * @deprecated Prior to 2.4 this method was the standard way to access {@link Metric}s by their {@link MetricID};
+     *             Doing so might cause unnecessary creation of an intermediate map that might be computed by the 
+     *             implementation. Prefer {@link #getMetric(MetricID)} for key based lookup of single {@link Metric};
+     *             Use {@link #getMetricIDs()} to iterate over all metrics. 
+     *             Use {@link #getMetrics(MetricFilter)} with {@link MetricFilter#ALL} if a (most likely) computed 
+     *             {@link Map} is preferred.
```

Comment:
Because it is obvious to the caller, that the result is almost certainly a computed map. That there is work to be done to receive it and that it is stable and disconnected from any changes to the registry. `getMetrics()` on the other hand might make you think it is as good as using `getMetric(String)` as it is `getMetrics().get(String)` which it is most certainly not. This is especially problematic since accessing single metrics by using the map has been the only option so far when doing lookup by `MetricID` so there is a high likelihood that users keep going that known path.